src/layout.rs

```diff
@@ -284,3 +287,221 @@ where
+            rescale(&mut pos, scale, (0..node_num).collect());
+        }
+    }
+
+    if let Some(center) = center {
+        recenter(&mut pos, center);
+    }
+
+    Pos2DMapping {
+        pos_map: graph.node_indices().map(|n| n.index()).zip(pos).collect(),
+    }
+}
+
+pub fn shell_layout<Ty: EdgeType>(
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    nlist: Option<Vec<Vec<usize>>>,
+    rotate: Option<f64>,
+    scale: Option<f64>,
+    center: Option<Point>,
+) -> Pos2DMapping {
+    let node_num = graph.node_bound();
+    let mut pos: Vec<Point> = vec![[0.0, 0.0]; node_num];
+    let pi = std::f64::consts::PI;
+
+    let shell_list: Vec<Vec<usize>> = match nlist {
+        Some(nlist) => nlist,
+        None => vec![graph.node_indices().map(|n| n.index()).collect()],
+    };
+    let shell_num = shell_list.len();
+
+    let radius_bump = match scale {
+        Some(scale) => scale / shell_num as f64,
+        None => 1.0 / shell_num as f64,
+    };
+
+    let mut radius = match node_num {
+        1 => 0.0,
+        _ => radius_bump,
+    };
+
+    let rot_angle = match rotate {
+        Some(rotate) => rotate,
+        None => pi / shell_num as f64,
+    };
+
+    let mut first_theta = rot_angle;
+    for shell in shell_list {
+        let shell_len = shell.len();
+        for i in 0..shell_len {
+            let angle = 2.0 * pi * i as f64 / shell_len as f64 + first_theta;
+            pos[shell[i]] = [radius * angle.cos(), radius * angle.sin()];
+        }
+        radius += radius_bump;
+        first_theta += rot_angle;
+    }
+
+    if let Some(center) = center {
+        recenter(&mut pos, center);
+    }
+
+    Pos2DMapping {
+        pos_map: graph
+            .node_indices()
+            .map(|n| {
+                let n = n.index();
+                (n, pos[n])
+            })
+            .collect(),
+    }
+}
+
+pub fn spiral_layout<Ty: EdgeType>(
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    scale: Option<f64>,
+    center: Option<Point>,
+    resolution: Option<f64>,
+    equidistant: Option<bool>,
+) -> Pos2DMapping {
+    let node_num = graph.node_count();
+    let mut pos: Vec<Point> = Vec::with_capacity(node_num);
+
+    let ros = resolution.unwrap_or(0.35);
+
+    if node_num == 1 {
+        pos.push([0.0, 0.0]);
+    } else if equidistant == Some(true) {
+        let mut theta: f64 = ros;
+        let chord = 1.0;
+        let step = 0.5;
+        for _ in 0..node_num {
+            let r = step * theta;
+            theta += chord / r;
+            pos.push([theta.cos() * r, theta.sin() * r]);
+        }
+    } else {
+        let mut angle: f64 = 0.0;
+        let mut dist = 0.0;
+        let step = 1.0;
+        for _ in 0..node_num {
+            pos.push([dist * angle.cos(), dist * angle.sin()]);
+            dist += step;
+            angle += ros;
+        }
+
+        if let Some(scale) = scale {
+            rescale(&mut pos, scale, (0..node_num).collect());
+        }
+    }
+
```

Comment:
```suggestion
    }
    
    if let Some(scale) = scale {
        rescale(&mut pos, scale, (0..node_num).collect());
    }

```

Reply:
Nice catch, I forgot there's an `else if` -- but the problem of moving it out is that if the graph only has 1 node, `lim` in `rescale` will be `0.0` and the point will become `[NaN, NaN]`. I guess I can modify `rescale` so it only does something when `lim > 0`.